### ISSUE TITLE
Mysql model generation takes very long and exit with error

### ISSUE THREAD

**Issue by nqsang90** [volatiletech/sqlboiler#329]
If you're having a generation problem please answer these questions before submitting your issue. Thanks!

### What version of SQLBoiler are you using (`sqlboiler --version`)?
SQLBoiler v2.7.3

### If this happened at generation time what was the full SQLBoiler command you used to generate your models? (if not applicable leave blank)
`sqlboiler mysql`

### What is the output of the command above with the `-d` flag added to it? (Provided you are comfortable sharing this, it contains a blueprint of your schema)
Part of the exception is. It seems the connection takes so long and closes unexpectedly.
`packets.go:36: unexpected EOF (Invalid Connection)`
Error comes from here:
`github.com/volatiletech/sqlboiler/bdb/interface.go:62`

### Please provide a relevant database schema so we can replicate your issue (Provided you are comfortable sharing this)
Problem is this query takes very long to run
```
SELECT
	c.column_name,
	c.column_type,
	IF(c.data_type = 'enum', c.column_type, c.data_type),
	IF(extra = 'auto_increment','auto_increment', c.column_default),
	c.is_nullable = 'YES',
		EXISTS (
			SELECT c.column_name
			FROM information_schema.table_constraints tc
			INNER JOIN information_schema.key_column_usage kcu
				ON tc.constraint_name = kcu.constraint_name AND tc.table_name = kcu.table_name AND tc.table_schema = kcu.table_schema
			WHERE c.column_name = kcu.column_name AND tc.table_name = c.table_name AND
				(tc.constraint_type = 'PRIMARY KEY' OR tc.constraint_type = 'UNIQUE') AND
				(SELECT count(*) FROM information_schema.key_column_usage WHERE table_schema = kcu.table_schema AND TABLE_NAME = tc.table_name AND CONSTRAINT_NAME = tc.constraint_name) = 1
		) AS is_unique
	FROM information_schema.columns AS c
	WHERE TABLE_NAME = 'table_name_here' AND table_schema = 'schema_name_here' AND c.extra NOT LIKE '%VIRTUAL%';
```

Explain the query:
```
id	select_type	table	type	possible_keys	key	key_len	ref	rows	Extra
1	PRIMARY	c	ALL	NULL	TABLE_SCHEMA,TABLE_NAME	NULL	NULL	NULL	Using where; Open_frm_only; Scanned 0 databases
2	DEPENDENT SUBQUERY	tc	ALL	NULL	NULL	NULL	NULL	NULL	Using where; Open_full_table; Scanned all databases
2	DEPENDENT SUBQUERY	kcu	ALL	NULL	NULL	NULL	NULL	NULL	Using where; Open_full_table; Scanned all databases; Using join buffer (Block Nested Loop)
3	DEPENDENT SUBQUERY	key_column_usage	ALL	NULL	NULL	NULL	NULL	NULL	Using where; Open_full_table; Scanned all databases
```

I wish someone can help with this.

**Comment by aarondl** [comment]
I've always noticed that mysql generation was slow. How big is your database that this is happening to you? Also what's in the mysql logs when this happens?

**Comment by nqsang90** [comment]
I can hardly tell this. I think it's not the database itself is big, but the query seems to scan all the databases. Because when I try to generate models for a small database, it still takes long. 
Using beego generation command on the same database takes about 1-2 minute. Maybe you can have some hint by looking into beego code. :)

**Comment by oderwat** [comment]
I have to chime in with a similar problem. The above mentioned query is needing 173 seconds for just one table in my case. But there are over 200 tables in this database. So it seems to need about 9-10 hours to complete running those queries alone. 

For info: Those 200 (208 exactly) tables can have up to 100 columns. This one database has  about 10.000.000 rows occupying about 5 GB data (including indexes). It is a pretty strongly normalised and completely handcrafted database (no ORM or other tools).

All in all there are around 4000 tables spread over all of the 29 databases on my development server. They occupy about 50 gigabyte of data total.

The number of 173 seconds was obtained by watching the processes and copying the below query and calling it with just one of the tables in the schema like this:

```
select
	c.column_name,
	c.column_type,
	if(c.data_type = 'enum', c.column_type, c.data_type),
	if(extra = 'auto_increment','auto_increment', c.column_default),
	c.is_nullable = 'YES',
		exists (
			select c.column_name
			from information_schema.table_constraints tc
			inner join information_schema.key_column_usage kcu
				on tc.constraint_name = kcu.constraint_name and tc.table_name = kcu.table_name and tc.table_schema = kcu.table_schema
			where c.table_name = tc.table_name and c.column_name = kcu.column_name and c.table_schema = kcu.constraint_schema and 
				(tc.constraint_type = 'PRIMARY KEY' or tc.constraint_type = 'UNIQUE') and
				(select count(*) from information_schema.key_column_usage where table_schema = kcu.table_schema and constraint_schema = kcu.table_schema and table_name = tc.table_name and constraint_name = tc.constraint_name) = 1
		) as is_unique
	from information_schema.columns as c
	where table_name = "users" and table_schema = "projectone" and c.extra not like '%VIRTUAL%' order by c.ordinal_position
```

I could successfully run sqlboiler on a smaller database on the same server containing four different tables. It still was a pretty slow process.

```
> time sqlboiler mysql

real	2m9.858s
user	0m0.142s
sys	0m0.061s
```

This is a mariadb server which is setup to be very fast using a lot of ram and residing on a thunderbolt ssd drive. That database does multiple thousands of queries per microsecond usually. So it is by no means a slow database or some small laptop system.

I guess the query itself explodes because of the many tables across all the databases on that server. But what I want to do is actually using SQLBoiler for such a project in the future. Starting with a smaller one but if I can't even generate the code in a decent timeframe, it does not make any sense to follow this path.

**Comment by aarondl** [comment]
Hi @oderwat. It definitely seems like you're getting to the crux of the issue here.

Perhaps there is some query optimization we can do here but it seems a bit messy to get into temporary tables. Also if we were to create temporary tables it would hardly make sense to delete them as you'd need them again the very next run. All sqlboiler drivers are unique and their own fully-fledged programs as of version 3. I wonder if standard optimization could be made to work here, replace the subselect with some joins, stare at `EXPLAIN` see if we can't join in an order that reduces the number of columns before the result set balloons.

It's unfortunate because without generation being fast sqlboiler really loses its charm as a tool.

More unfortunately I'm not interested in pursuing the fix for this (exclusively postgresql here :p) so it's up to the sqlboiler community of mysql users to figure this one out! I'd be happy to shepherd a PR to fix this in.

Best of luck if you do attempt to solve this, and if you don't then I can't blame you!

**Comment by oderwat** [comment]
I am not sure about joins as I fear the main problem here is that there are no keys at all in the views for this tables. Those are read only too, so this seems not to be changeable. This is where the temporary tables idea came from. By simply copying the views into tables, adding an index and using those. This should be really fast and I guess that would solve the problem quite easily.

We may want to search for a solution and doing a PR. Same goes for the MariaDB problem with the SSL differences to MySQL. While this seems to be more a problem upstream actually.

There is also another problem we may run into as we usually use german names for tables which makes it hard to guesstimate the plural and non plural names. I actually fear that this is a bigger problem too. Thats one of the reasons I wanted to start with converting some of the mysql projects we habe.

I had an attempt to change and compile sqlboiler-mysql but there goes something on, that I don't understand. I can't get my version to be used in generation. I thought changing the driver files, building them and copying it to GOHOME/bin should work. It seems that is not the case and I am missing something fundamental here. But I didn't dig deeper into it.

I find it refreshening that you say you care for postgresql only. For two reasons: I want to use postgresql for the project I validate sqlboilder for. Second it does not feel like I would do something you would do in a snap, when I try to solve the mysql problem :)

**Comment by aarondl** [comment]
When you execute sqlboiler, you can give a full path to the driver you'd like to execute. Usually what I do when developing is:

```sqlboiler --wipe ./sqlboiler-psql```

Also I wonder if we couldn't cache some data using tmp files in the mysql driver rather than in temporary tables. That way permissions can stay slim for generation. I think it already uses tmp files for something for mysql.

**Comment by oderwat** [comment]
OK. I got some minimal progress after finding out that build all does not build all. Then I stumbled of an older version of go-bindata on my system. Well, I can compile and run my stuff!

First thing I encountered is, that all tests for my little schema fail. This is because there is an unique index. The randomizer does not generate a random string in non special string fields, therefor it tries to insert always an empty string, which of course is not unique.

I am also not yet sure if sqlboiler actually "knows" what it means when there is an unique index for any column? I saw something about a problem with unique foreign keys when they are using a composition of multiple columns. But I don't want to waste your time until I understand more of whats going on.

So I am going to make some changes until I can use it to actually validate how it could server our needs.

You maybe want to open this issue again?

**Comment by aarondl** [comment]
Very happy to open it if someone's working on it!

**Comment by aarondl** [comment]
> First thing I encountered is, that all tests for my little schema fail. This is because there is an unique index. The randomizer does not generate a random string in non special string fields, therefor it tries to insert always an empty string, which of course is not unique.

This is very true. The randomized testing is actually somewhat impossible to do correctly without knowledge of the schema because the schema has many rules the randomizer cannot be aware of without a lot of work (like enum restrictions, or a char_length(col) > 0 check).

sqlboiler itself doesn't need to understand unique indexes, just the tests (and they ignore them currently).

**Comment by oderwat** [comment]
OK. I also had the case that testing failed because a unique index existed for a column which has a default value. This is of course nonsense and I can't blame sqlboiler really.

We are using MyISAM tables for performance reasons. So the main problem with the tests resulted from them using transactions which is not possible with MyISAM. It think the test suite should bail out if that is the case. I converted my little test db to InnoDB so I could continue.

The next thing which caught me is that there seems to be no support for the "SET" datatype or at least the tests don't work for the tables which have tests (truncation errors on insert). Could your shortly confirm if SETs are supported or not?

I thought it may be possible to generate the schemas and use the type checker abstraction for writing (better) queries, even when the database was not designed to work with sqlboiler. Maybe that was naive to think in the first place.

We may just skip the tests for now and try do do something about the generation time. I am at least more confident about modifying the code.

**Comment by aarondl** [comment]
I don't have any clue about the `SET` datatype in mysql. If it doesn't behave like everything else then it's probably not handled at all.

I'm marking this closed as I've merged #732 into the dev branch (it will show closed because it was opened against the wrong branch).

**Comment by oderwat** [comment]
No problem. I will investigate the SET problematic later as we use it with some schemas. Thank you!
